File: src/main/java/org/cimsbioko/repository/Query.java
package org.cimsbioko.repository;

import android.database.Cursor;
import android.net.Uri;

import static org.cimsbioko.App.getApp;
import static org.cimsbioko.repository.RepositoryUtils.EQUALS;
import static org.cimsbioko.repository.RepositoryUtils.buildWhereStatement;

/**
 * Represent a database query to be performed.  Might be saved and performed in pieces by an Iterator.
 */
public class Query {

    private final Uri tableUri;
    private final String[] columnNames;
    private final String[] columnValues;
    private final String columnOrderBy;
    private final String operator;

    // simple query on one column equals value
    public Query(Uri tableUri, String columnName, String columnValue, String columnOrderBy) {
        this.tableUri = tableUri;
        this.columnNames = null == columnName ? null : new String[] {columnName};
        this.columnValues = null == columnValue ? null : new String[] {columnValue};
        this.columnOrderBy = columnOrderBy;
        this.operator = EQUALS;
    }

    // full, flexible query
    public Query(Uri tableUri, String[] columnNames, String[] columnValues, String columnOrderBy, String operator) {
        this.tableUri = tableUri;
        this.columnNames = columnNames;
        this.columnValues = columnValues;
        this.columnOrderBy = columnOrderBy;
        this.operator = operator;
    }

    public Cursor select() {
        return getApp()
                .getContentResolver()
                .query(tableUri, null, buildWhereStatement(columnNames, operator), columnValues, columnOrderBy);
    }
}

File: src/main/kotlin/org/cimsbioko/data/Query.kt
package org.cimsbioko.data

import android.database.Cursor
import android.net.Uri
import org.cimsbioko.App
import org.cimsbioko.data.CursorConvert.list
import org.cimsbioko.data.CursorConvert.one
import org.cimsbioko.data.WhereUtils.buildWhereStatement
import org.cimsbioko.navconfig.UsedByJSConfig

const val EQUALS = "="
const val LIKE = "LIKE"

private const val AND = "AND"
private const val WHERE_PLACEHOLDER = "?"
private const val WHERE_ALL = "1"

/**
 * Represent a database query to be performed.  Might be saved and performed in pieces by an Iterator.
 */
class Query<T : Any> internal constructor(
        private val gateway: Gateway<T>,
        private val tableUri: Uri,
        columnName: String?,
        columnValue: String?,
        private val columnOrderBy: String,
        private val operator: String = EQUALS) {

    private val columnNames: Array<String>? = columnName?.let { arrayOf(it) }
    private val columnValues: Array<String>? = columnValue?.let { arrayOf(it) }

    private fun select(): Cursor? {
        return App.instance
                .contentResolver
                .query(tableUri, null, buildWhereStatement(columnNames ?: emptyArray(), operator), columnValues, columnOrderBy)
    }

    val first: T?
        get() = one(select(), gateway.entityConverter)

    @get:UsedByJSConfig
    val list: List<T>
        get() = list(select(), gateway.entityConverter)

    // get the first result from a query as a QueryResult or null
    val firstWrapper: DataWrapper?
        get() = one(select(), gateway.wrapperConverter)

    // get all results from a query as a list of QueryResults
    val wrapperList: List<DataWrapper>
        get() = list(select(), gateway.wrapperConverter)

    fun exists(): Boolean {
        return select()?.moveToFirst() == true
    }
}

internal object WhereUtils {

    fun buildWhereStatement(columnNames: Array<String>, operator: String): String =
            if (columnNames.isNotEmpty()) {
                fun buildWhereClause(column: String, op: String) = "$column $op $WHERE_PLACEHOLDER"
                buildString {
                    columnNames.firstOrNull()?.let { first ->
                        append(buildWhereClause(first, operator))
                        columnNames.drop(1).forEach {
                            append(" $AND ")
                            append(buildWhereClause(it, operator))
                        }
                    }
                }
            } else WHERE_ALL

}